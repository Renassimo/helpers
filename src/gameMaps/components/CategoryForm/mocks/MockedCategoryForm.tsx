const MockedCategoryForm = jest.fn(({ values, errors, isReady }) => (
  <div>
    MockedCategoryForm - {values?.title} - {values?.description} -{' '}
    {values?.color} - {values?.itemsAmount} - {errors?.title} -{' '}
    {errors?.description} - {errors?.color} - {errors?.itemsAmount} -{' '}
    {isReady && 'isReady'}
  </div>
));

export default MockedCategoryForm;
