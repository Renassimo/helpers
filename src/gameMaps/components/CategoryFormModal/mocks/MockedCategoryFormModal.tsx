const MockedCategoryFormModal = jest.fn(({ isModalOpen, gameId }) => (
  <div>
    MockedCategoryFormModal {isModalOpen ? 'opened' : 'closed'} - {gameId}
  </div>
));

export default MockedCategoryFormModal;
