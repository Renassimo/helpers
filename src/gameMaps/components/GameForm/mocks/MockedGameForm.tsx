const MockedGameForm = jest.fn(({ values, errors }) => (
  <div>
    MockedGameForm - {values?.title} - {values?.description} -{' '}
    {values?.backgroundColor} - {values?.mapImageUrl} - {errors?.title} -{' '}
    {errors?.description} - {errors?.backgroundColor}
  </div>
));

export default MockedGameForm;
