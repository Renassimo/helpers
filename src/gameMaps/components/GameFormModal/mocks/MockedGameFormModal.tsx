const MockedGameFormModal = jest.fn(({ isModalOpen }) => (
  <div>MockedGameFormModal {isModalOpen ? 'opened' : 'closed'}</div>
));

export default MockedGameFormModal;
