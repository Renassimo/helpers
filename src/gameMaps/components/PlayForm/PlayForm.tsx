import TextField from '@mui/material/TextField';
import Alert from '@mui/material/Alert';
import Button from '@mui/material/Button';

import { PlayFormProps } from '@/gameMaps/types/props';

const PlayForm = ({
  values,
  setters,
  errors = {},
  onDelete,
}: PlayFormProps) => {
  const { title, description } = values;
  const { setTitle, setDescription } = setters;

  return (
    <>
      <TextField
        name="title"
        label="Title"
        value={title}
        onChange={(event) => setTitle(event.target.value)}
        fullWidth
        margin="dense"
        variant="standard"
        error={!!errors.title}
        helperText={errors.title}
      />
      <TextField
        name="description"
        label="Description"
        value={description}
        onChange={(event) => setDescription(event.target.value)}
        fullWidth
        multiline
        margin="dense"
        variant="standard"
        error={!!errors.description}
        helperText={errors.description}
      />
      {onDelete && (
        <Button
          type="button"
          fullWidth
          variant="outlined"
          onClick={onDelete}
          color="error"
          sx={{ marginTop: 3 }}
        >
          Delete
        </Button>
      )}
      {errors.main && <Alert severity="error">{errors.main}</Alert>}
    </>
  );
};

export default PlayForm;
