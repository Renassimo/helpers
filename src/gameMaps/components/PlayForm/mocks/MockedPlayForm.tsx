const MockedPlayForm = jest.fn(({ values, errors, isReady }) => (
  <div>
    MockedPlayForm - {values?.title} - {values?.description} - {errors?.title} -{' '}
    {errors?.description} - {isReady && 'isReady'}
  </div>
));

export default MockedPlayForm;
