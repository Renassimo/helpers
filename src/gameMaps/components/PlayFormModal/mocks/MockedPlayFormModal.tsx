const MockedPlayFormModal = jest.fn(({ isModalOpen }) => (
  <div>MockedPlayFormModal {isModalOpen ? 'opened' : 'closed'}</div>
));

export default MockedPlayFormModal;
