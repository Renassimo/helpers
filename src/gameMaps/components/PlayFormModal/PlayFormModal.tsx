import { useEffect } from 'react';

import usePlayForm from '@/gameMaps/hooks/usePlayForm';

import Modal from '@/common/components/Modal';
import PlayForm from '@/gameMaps/components/PlayForm';

import { PlayData } from '@/gameMaps/types';

const PlayFormModal = ({
  isModalOpen,
  setIsModalOpen,
  data,
  onFinish,
  gameId,
}: {
  isModalOpen: boolean;
  setIsModalOpen: (isOpen: boolean) => void;
  data?: PlayData | null;
  onFinish: (newData: PlayData | null) => void;
  gameId: string;
}) => {
  const onModalClose = () => {
    setIsModalOpen(false);
    cleanForm();
  };

  const handleFinish = (newData: PlayData | null) => {
    onFinish(newData);
    onModalClose();
  };

  const {
    isEditForm,
    values,
    setters,
    cleanForm,
    prepareFormForEdit,
    onSubmit,
    onDelete,
    errors,
    loading,
  } = usePlayForm(gameId, data, handleFinish);

  const handleDelete = async (): Promise<void> => {
    if (
      confirm(
        `Are you sure to delete "${
          data?.attributes.title ?? values.title
        }" play? Following action will also delete all found items related to this play.`
      )
    ) {
      await onDelete();
      handleFinish(null);
    }
  };

  useEffect(() => {
    if (isModalOpen && isEditForm) {
      prepareFormForEdit();
    }
  }, [isModalOpen, isEditForm]);

  return (
    <Modal
      open={isModalOpen}
      onClose={onModalClose}
      onSubmit={onSubmit}
      title={`${isEditForm ? 'Update' : `Create new`} play`}
      loading={loading}
    >
      <PlayForm
        values={values}
        setters={setters}
        errors={errors}
        onDelete={isEditForm ? handleDelete : undefined}
      />
    </Modal>
  );
};

export default PlayFormModal;
