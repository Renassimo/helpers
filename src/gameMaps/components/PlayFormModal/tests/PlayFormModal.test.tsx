import renderWithTheme from '@/common/tests/helpers/renderWithTheme';

import usePlayForm from '@/gameMaps/hooks/usePlayForm';

import Modal from '@/common/components/Modal';
import PlayForm from '@/gameMaps/components/PlayForm';

import MockedModal from '@/common/components/Modal/mocks';
import MockedPlayForm from '@/gameMaps/components/PlayForm/mocks';
import { mockedGame, mockedPlay } from '@/gameMaps/types/mocks';

import PlayFormModal from '../PlayFormModal';

jest.mock('@/gameMaps/hooks/usePlayForm');
jest.mock('@/common/components/Modal');
jest.mock('@/gameMaps/components/PlayForm');

describe('PlayFormModal', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    (Modal as unknown as jest.Mock).mockImplementation(MockedModal);
    (PlayForm as unknown as jest.Mock).mockImplementation(MockedPlayForm);
  });

  const mockedData = mockedPlay;
  const mockedGameId = mockedGame.id;

  const mockedValues = {
    title: 'mocked-title',
    description: 'mocked-description',
  };
  const mockedSetters = 'mocked-setters';
  const mockedCleanForm = jest.fn();
  const mockedPrepareFormForEdit = jest.fn();
  const mockedOnSubmit = jest.fn();
  const mockedSetIsModalOpen = jest.fn();
  const mockedErrors = 'mockedErrors';
  const mockedOnFinish = jest.fn();

  test('renders snapshot successfully', () => {
    // Arange
    const mockedUsePlayForm = jest.fn(() => ({
      isEditForm: false,
      values: mockedValues,
      setters: mockedSetters,
      cleanForm: mockedCleanForm,
      prepareFormForEdit: mockedPrepareFormForEdit,
      onSubmit: mockedOnSubmit,
      errors: mockedErrors,
    }));
    (usePlayForm as unknown as jest.Mock).mockImplementation(mockedUsePlayForm);
    // Act
    const { baseElement } = renderWithTheme(
      <PlayFormModal
        isModalOpen={true}
        setIsModalOpen={mockedSetIsModalOpen}
        onFinish={mockedOnFinish}
        gameId={mockedGameId}
      />
    );
    // Assert
    expect(baseElement).toMatchSnapshot();
    expect(mockedUsePlayForm).toBeCalledWith(
      mockedGameId,
      undefined,
      expect.any(Function)
    );
    expect(mockedPrepareFormForEdit).not.toBeCalled();
    expect(MockedPlayForm).toHaveBeenCalledWith(
      { values: mockedValues, setters: mockedSetters, errors: mockedErrors },
      {}
    );
  });

  describe('when data passed', () => {
    test('renders snapshot successfully', async () => {
      // Arange
      const mockedUsePlayForm = jest.fn(() => ({
        isEditForm: true,
        values: mockedValues,
        setters: mockedSetters,
        cleanForm: mockedCleanForm,
        prepareFormForEdit: mockedPrepareFormForEdit,
        onSubmit: mockedOnSubmit,
        errors: mockedErrors,
      }));
      (usePlayForm as unknown as jest.Mock).mockImplementation(
        mockedUsePlayForm
      );
      // Act
      const { baseElement } = await renderWithTheme(
        <PlayFormModal
          isModalOpen={true}
          setIsModalOpen={mockedSetIsModalOpen}
          onFinish={mockedOnFinish}
          data={mockedData}
          gameId={mockedGameId}
        />
      );
      // Assert
      expect(baseElement).toMatchSnapshot();
      expect(mockedUsePlayForm).toBeCalledWith(
        mockedGameId,
        mockedData,
        expect.any(Function)
      );
      expect(mockedPrepareFormForEdit).toBeCalled();
      expect(MockedPlayForm).toHaveBeenCalledWith(
        {
          values: mockedValues,
          setters: mockedSetters,
          errors: mockedErrors,
          onDelete: expect.any(Function),
        },
        {}
      );
    });

    test('calls prepareFormForEdit', () => {
      // Arange
      const mockedUsePlayForm = jest.fn(() => ({
        isEditForm: true,
        values: mockedValues,
        setters: mockedSetters,
        cleanForm: mockedCleanForm,
        prepareFormForEdit: mockedPrepareFormForEdit,
        onSubmit: mockedOnSubmit,
        errors: mockedErrors,
      }));
      (usePlayForm as unknown as jest.Mock).mockImplementation(
        mockedUsePlayForm
      );
      // Act
      renderWithTheme(
        <PlayFormModal
          isModalOpen={true}
          setIsModalOpen={mockedSetIsModalOpen}
          data={mockedData}
          onFinish={mockedOnFinish}
          gameId={mockedGameId}
        />
      );
      // Assert
      expect(mockedUsePlayForm).toBeCalledWith(
        mockedGameId,
        mockedData,
        expect.any(Function)
      );
      expect(mockedPrepareFormForEdit).toBeCalled();
    });
  });
});
