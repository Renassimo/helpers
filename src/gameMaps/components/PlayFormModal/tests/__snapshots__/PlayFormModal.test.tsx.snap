// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayFormModal renders snapshot successfully 1`] = `
<body>
  <div>
    <div>
      <div>
        MockedPlayForm - 
        mocked-title
         - 
        mocked-description
         - 
         -
         
      </div>
      <div>
        title: 
        Update play
        ; maxWidth: 
        sm
        ;
        open
        ;
        not loading
        ;
        not disabled
        ; onClose
         
        passed
        ; onSubmit
         
        passed
        ;
      </div>
    </div>
  </div>
</body>
`;

exports[`PlayFormModal when data passed renders snapshot successfully 1`] = `
<body>
  <div>
    <div>
      <div>
        MockedPlayForm - 
        mocked-title
         - 
        mocked-description
         - 
         -
         
      </div>
      <div>
        title: 
        Create new play
        ; maxWidth: 
        sm
        ;
        open
        ;
        not loading
        ;
        not disabled
        ; onClose
         
        passed
        ; onSubmit
         
        passed
        ;
      </div>
    </div>
  </div>
</body>
`;
